self-core-impl/src/main/java/com/selfxdsd/core/projects/FakeWallet.java

```diff
@@ -0,0 +1,257 @@
+     */
+    private final boolean active;
+
+    /**
+     * Wallet ID.
+     */
+    private final String identifier;
+
+    /**
+     * Ctor.
+     * @param storage Self storage.
+     * @param project Project to which this wallet belongs.
+     * @param limit Cash limit we're allowed to use.
+     * @param identifier Wallet identifier.
+     * @param active Is this wallet active or not?
+     */
+    public FakeWallet(
+        final Storage storage,
+        final Project project,
+        final BigDecimal limit,
+        final String identifier,
+        final boolean active
+    ) {
+        this.storage = storage;
+        this.project = project;
+        this.limit = limit;
+        this.identifier = identifier;
+        this.active = active;
+    }
+
+    @Override
+    public BigDecimal cash() {
+        return this.limit;
+    }
+
+    @Override
+    public Wallet pay(final Invoice invoice) {
+        if (!this.project.equals(invoice.contract().project())) {
+            throw new InvoiceException.NotPartOfProjectContract(invoice,
+                this.project);
+        }
+        if(invoice.isPaid()) {
+            throw new InvoiceException.AlreadyPaid(invoice);
+        }
+        final BigDecimal newCash = this.limit.subtract(invoice
+            .totalAmount());
+        if (newCash.longValueExact() < 0L) {
+            throw new WalletPaymentException("No cash available in wallet "
+                + "for paying invoice #" + invoice.invoiceId()
+                + ". Please increase the limit from your dashboard with"
+                + " at least " + newCash.abs().divide(BigDecimal
+                .valueOf(1000), RoundingMode.HALF_UP) + "\u20ac."
+            );
+        }
+        LOG.debug(
+            "[FAKE] Paying Invoice #" + invoice.invoiceId()
+                + " of Contract " + invoice.contract().contractId()
+                + "..."
+        );
+        final boolean paid = this.storage
+            .invoices()
+            .registerAsPaid(new StoredInvoice(
+                invoice.invoiceId(),
+                invoice.contract(),
+                invoice.createdAt(),
+                LocalDateTime.now(),
+                UUID.randomUUID().toString().replace("-", ""),
+                this.storage
+            ));
+        if (!paid) {
+            throw new WalletPaymentException(
+                "Could not pay invoice #" + invoice.invoiceId()
+            );
+        }
+        return new FakeWallet(
+            this.storage,
+            this.project,
+            newCash,
+            this.identifier, this.active
+        );
+    }
+
+    @Override
+    public String type() {
+        return Type.FAKE;
+    }
+
+    @Override
+    public boolean active() {
+        return this.active;
+    }
+
+    @Override
+    public Project project() {
+        return this.project;
+    }
+
+    @Override
+    public Wallet updateCash(final BigDecimal cash) {
+        return new FakeWallet(
```

Comment:
@amihaiemil should we update this to storage instead? (though current storage impl forbids fake wallet to update the cash)

Reply:
@criske yes, it needs to be updated in the storage too, otherwise users will be able to use the FakeWallet forever. We want to let them use it only up to the initial limit. Write a puzzle for that pls :D